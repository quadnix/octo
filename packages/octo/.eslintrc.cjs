module.exports = {
  extends: ['../../.eslintrc.cjs'],
  ignorePatterns: ['coverage', 'dist', 'node_modules'],
  parserOptions: {
    ecmaVersion: '2022',
    extraFileExtensions: ['.json'],
    project: 'tsconfig.json',
    sourceType: 'module',
    tsconfigRootDir: __dirname,
  },
  root: true,
  rules: {
    '@nx/enforce-module-boundaries': [
      'error',
      {
        allow: [],
        depConstraints: [
          {
            onlyDependOnLibsWithTags: ['scope:octo'],
            sourceTag: 'scope:octo',
          },
        ],
      },
    ],
    'boundaries/element-types': [
      'error',
      {
        default: 'allow',
        rules: [],
      },
    ],
    'boundaries/external': ['off'],
  },
  settings: {
    'boundaries/elements': [
      {
        pattern: 'src',
        type: 'src',
      },
    ],
    'import/resolver': {
      typescript: {
        alwaysTryTypes: true,
      },
    },
  },
};
