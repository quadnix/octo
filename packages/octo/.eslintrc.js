module.exports = {
  extends: ['../../.eslintrc.js'],
  ignorePatterns: ['coverage', 'dist', 'node_modules'],
  parserOptions: {
    extraFileExtensions: ['.json'],
    project: 'tsconfig.json',
    sourceType: 'module',
    tsconfigRootDir: __dirname,
  },
  root: true,
  rules: {
    'import/no-restricted-paths': [
      'error',
      {
        zones: [
          {
            except: ['./octo'],
            from: '..',
            target: '.',
          },
        ],
      },
    ],
  },
  settings: {
    // Fix to make `no-restricted-paths` rule work.
    // Source: https://github.com/import-js/eslint-plugin-import/issues/1928#issuecomment-715164532
    'import/resolver': {
      node: {
        extensions: ['.ts'],
      },
    },
  },
};
