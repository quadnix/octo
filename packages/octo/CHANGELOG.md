## 0.0.4 (2024-04-02)

This was a version bump only for @quadnix/octo to align it with other projects, there were no code changes.