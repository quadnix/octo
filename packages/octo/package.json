{
  "dependencies": {
    "reflect-metadata": "~0.1.13",
    "typedi": "~0.10.0"
  },
  "description": "Octo is an opinionated interface of a cloud infrastructure.",
  "devDependencies": {},
  "engines": {
    "node": ">=16"
  },
  "exports": "./dist/index.js",
  "jest": {
    "coverageDirectory": "./coverage",
    "extensionsToTreatAsEsm": [
      ".ts"
    ],
    "moduleFileExtensions": [
      "js",
      "json",
      "ts"
    ],
    "moduleNameMapper": {
      "(.+)\\.js": "$1"
    },
    "preset": "ts-jest/presets/default-esm",
    "rootDir": ".",
    "testEnvironment": "node",
    "testRegex": ".*spec\\.ts$",
    "transform": {
      "^.+\\.ts$": [
        "ts-jest",
        {
          "useESM": true
        }
      ]
    }
  },
  "name": "@quadnix/octo",
  "nx": {
    "tags": [
      "scope:octo"
    ]
  },
  "scripts": {
    "build": "rimraf dist && tsc -p tsconfig.build.json",
    "format": "prettier --write \"src/**/*.ts\" \"test/**/*.ts\"",
    "lint": "eslint .",
    "lint:fix": "eslint \"{src,apps,libs,test}/**/*.ts\" --fix",
    "test": "NODE_OPTIONS=--experimental-vm-modules jest --testRegex=\"\\.spec.ts$\"",
    "test:cov": "jest --coverage --testRegex=\"\\.spec.ts$\"",
    "test:debug": "node --inspect-brk -r tsconfig-paths/register -r ts-node/register node_modules/.bin/jest --runInBand",
    "test:e2e": "jest --testRegex=\"\\.e2e-spec.ts$\"",
    "test:integration": "jest --testRegex=\"\\.ispec.ts$\"",
    "test:watch": "jest --watch"
  },
  "type": "module",
  "version": "0.0.1"
}
