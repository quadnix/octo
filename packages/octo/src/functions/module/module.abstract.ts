import { Model } from '../../models/model.abstract';
import { IModule } from './module.interface';

export abstract class Module {
  private readonly members: Model<unknown, unknown>[] = [];

  protected addMember(member: Model<unknown, unknown>): void {
    const index = this.members.findIndex((m) => m.getContext() === member.getContext());

    if (index === -1) {
      this.members.push(member);
    }
  }

  abstract apply(...args: Model<unknown, unknown>[]): void;

  abstract remove(...args: Model<unknown, unknown>[]): void;

  removeAll(): void {
    for (const member of this.members) {
      member.remove(false, true);
    }

    for (const member of this.members) {
      member.remove(false, false);
    }
  }

  protected removeMember(member: Model<unknown, unknown>): void {
    const index = this.members.findIndex((m) => m.getContext() === member.getContext());

    if (index !== -1) {
      this.members.splice(index, 1);
    }
  }

  synth(): IModule {
    return {
      members: this.members.map((m) => ({
        context: m.getContext(),
      })),
    };
  }

  static async unSynth(
    deserializationClass: any,
    module: IModule,
    deReferenceContext: (context: string) => Promise<Model<unknown, unknown>>,
  ): Promise<Module> {
    const newModule = new deserializationClass();

    for (const member of module.members) {
      const model = await deReferenceContext(member.context);
      newModule.addMember(model);
    }

    return newModule;
  }
}
