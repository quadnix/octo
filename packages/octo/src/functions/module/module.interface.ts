import { IModelReference } from '../../models/model.interface';

export interface IModule {
  members: IModelReference[];
}
