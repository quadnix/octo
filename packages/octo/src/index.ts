export { Action } from './decorators/action.decorator.js';
export { Anchor } from './decorators/anchor.decorator.js';
export { Container } from './decorators/container.js';
export { EnableHook } from './decorators/enable-hook.decorator.js';
export { Factory } from './decorators/factory.decorator.js';
export { Model } from './decorators/model.decorator.js';
export { Module } from './decorators/module.decorator.js';
export { Overlay } from './decorators/overlay.decorator.js';
export { Resource } from './decorators/resource.decorator.js';
export { TestContainer } from './decorators/test-container.js';

export { DependencyRelationship } from './functions/dependency/dependency.js';
export { Diff, DiffAction } from './functions/diff/diff.js';
export { DiffUtility } from './functions/diff/diff.utility.js';
export { DiffMetadata } from './functions/diff/diff-metadata.js';

export { type IApp } from './models/app/app.interface.js';
export { App } from './models/app/app.model.js';
export { type IDeployment } from './models/deployment/deployment.interface.js';
export { Deployment } from './models/deployment/deployment.model.js';
export { type IEnvironment } from './models/environment/environment.interface.js';
export { Environment } from './models/environment/environment.model.js';
export { type IExecution } from './models/execution/execution.interface.js';
export { Execution } from './models/execution/execution.model.js';
export { type IImage } from './models/image/image.interface.js';
export { type IImageDockerOptions, Image } from './models/image/image.model.js';
export { type IPipeline } from './models/pipeline/pipeline.interface.js';
export { Pipeline } from './models/pipeline/pipeline.model.js';
export { type IRegion } from './models/region/region.interface.js';
export { Region } from './models/region/region.model.js';
export { type IServer } from './models/server/server.interface.js';
export { Server } from './models/server/server.model.js';
export { type IService } from './models/service/service.interface.js';
export { Service } from './models/service/service.model.js';
export { type ISubnet } from './models/subnet/subnet.interface.js';
export { Subnet, SubnetType } from './models/subnet/subnet.model.js';

export { AModel } from './models/model.abstract.js';
export { type IModel, type IModelReference } from './models/model.interface.js';
export { type IModelAction } from './models/model-action.interface.js';

export { AAnchor } from './overlays/anchor.abstract.js';
export { type IAnchor } from './overlays/anchor.interface.js';
export { AOverlay } from './overlays/overlay.abstract.js';
export { type IOverlay } from './overlays/overlay.interface.js';
export { OverlayService } from './overlays/overlay.service.js';

export { AResource } from './resources/resource.abstract.js';
export { type IResource } from './resources/resource.interface.js';
export { type IResourceAction } from './resources/resource-action.interface.js';
export { ASharedResource } from './resources/shared-resource.abstract.js';

export { InputService } from './services/input/input.service.js';

export { ModelSerializationService } from './services/serialization/model/model-serialization.service.js';
export { ResourceSerializationService } from './services/serialization/resource/resource-serialization.service.js';

export { LocalStateProvider } from './services/state-management/local.state-provider.js';
export { StateManagementService } from './services/state-management/state-management.service.js';
export { type IStateProvider } from './services/state-management/state-provider.interface.js';

export { TransactionService } from './services/transaction/transaction.service.js';

export * from './app.type.js';

export { Octo } from './main.js';
