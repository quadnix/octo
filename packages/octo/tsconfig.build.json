{
  "exclude": ["**/*spec.ts", "coverage", "dist", "node_modules", "test"],
  "extends": "./tsconfig.json"
}
