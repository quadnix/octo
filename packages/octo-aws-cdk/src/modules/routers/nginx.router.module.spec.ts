import { App, Environment, Image, ModelSerializationService, Region } from '@quadnix/octo';
import { NginxRouterModule } from './nginx.router.module';

describe('NginxRouterModule UT', () => {
  it('should test add and delete operations', async () => {
    const nginxRouterModule = new NginxRouterModule();
    const service = new ModelSerializationService();

    const app1 = new App('test');
    const nginxImage1 = new Image('quadnix/nginx', '0.0.1', {
      dockerFilePath: 'resources/images/quadnix/nginx/0.0.1',
    });
    app1.addImage(nginxImage1);
    const region1 = new Region('region-1');
    app1.addRegion(region1);
    const environment1 = new Environment('qa');
    region1.addEnvironment(environment1);

    // Creating router on environment, should create nginx router 0.0.1
    nginxRouterModule.apply(app1, nginxImage1, environment1);

    const diffs1 = await app1.diff();
    expect(diffs1).toMatchSnapshot();
    expect(nginxRouterModule.synth()).toMatchSnapshot();

    const app2 = (await service.deserialize(service.serialize(app1))) as App;
    const nginxImage2 = new Image('quadnix/nginx', '0.0.2', {
      dockerFilePath: 'resources/images/quadnix/nginx/0.0.2',
    });
    app2.addImage(nginxImage2);
    const region2 = app2.getChild('region', [{ key: 'regionId', value: 'region-1' }]) as Region;
    const environment2 = region2.getChild('environment', [{ key: 'environmentName', value: 'qa' }]) as Environment;

    // Creating router on environment, should create another nginx router 0.0.2 in the same module.
    nginxRouterModule.apply(app2, nginxImage2, environment2);

    const diffs2 = await app2.diff(app1);
    expect(diffs2).toMatchSnapshot();
    expect(nginxRouterModule.synth()).toMatchSnapshot();

    const app3 = (await service.deserialize(service.serialize(app2))) as App;
    const nginxImage3 = app3.getChild('image', [{ key: 'imageTag', value: '0.0.1' }]) as Image;
    const region3 = app3.getChild('region', [{ key: 'regionId', value: 'region-1' }]) as Region;
    const environment3 = region3.getChild('environment', [{ key: 'environmentName', value: 'qa' }]) as Environment;

    // Removing router 0.0.1 on environment, should remove nginx router 0.0.1 from module.
    nginxRouterModule.remove(app3, nginxImage3, environment3);

    const diffs3 = await app3.diff(app2);
    expect(diffs3).toMatchSnapshot();
    expect(nginxRouterModule.synth()).toMatchSnapshot();
  });
});
