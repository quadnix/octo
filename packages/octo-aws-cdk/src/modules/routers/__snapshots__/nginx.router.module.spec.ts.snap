// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NginxRouterModule UT should test add and delete operations 1`] = `
[
  {
    "action": "add",
    "field": "imageId",
    "value": "quadnix/nginx:0.0.1",
  },
  {
    "action": "add",
    "field": "regionId",
    "value": "region-1",
  },
  {
    "action": "add",
    "field": "environmentName",
    "value": "qa",
  },
  {
    "action": "add",
    "field": "serverKey",
    "value": "nginx",
  },
  {
    "action": "add",
    "field": "deploymentTag",
    "value": "nginx@0.0.1",
  },
  {
    "action": "add",
    "field": "executionId",
    "value": "nginx@0.0.1_qa",
  },
]
`;

exports[`NginxRouterModule UT should test add and delete operations 2`] = `
{
  "members": [
    {
      "context": "support=nginx,app=test",
    },
    {
      "context": "deployment=nginx@0.0.1,support=nginx,app=test",
    },
    {
      "context": "execution=nginx@0.0.1_qa,deployment=nginx@0.0.1,support=nginx,app=test,environment=qa,region=region-1,app=test",
    },
  ],
}
`;

exports[`NginxRouterModule UT should test add and delete operations 3`] = `
[
  {
    "action": "add",
    "field": "imageId",
    "value": "quadnix/nginx:0.0.2",
  },
  {
    "action": "add",
    "field": "deploymentTag",
    "value": "nginx@0.0.2",
  },
  {
    "action": "add",
    "field": "executionId",
    "value": "nginx@0.0.2_qa",
  },
]
`;

exports[`NginxRouterModule UT should test add and delete operations 4`] = `
{
  "members": [
    {
      "context": "support=nginx,app=test",
    },
    {
      "context": "deployment=nginx@0.0.1,support=nginx,app=test",
    },
    {
      "context": "execution=nginx@0.0.1_qa,deployment=nginx@0.0.1,support=nginx,app=test,environment=qa,region=region-1,app=test",
    },
    {
      "context": "deployment=nginx@0.0.2,support=nginx,app=test",
    },
    {
      "context": "execution=nginx@0.0.2_qa,deployment=nginx@0.0.2,support=nginx,app=test,environment=qa,region=region-1,app=test",
    },
  ],
}
`;

exports[`NginxRouterModule UT should test add and delete operations 5`] = `
[
  {
    "action": "delete",
    "field": "executionId",
    "value": "nginx@0.0.1_qa",
  },
  {
    "action": "delete",
    "field": "deploymentTag",
    "value": "nginx@0.0.1",
  },
]
`;

exports[`NginxRouterModule UT should test add and delete operations 6`] = `
{
  "members": [
    {
      "context": "support=nginx,app=test",
    },
    {
      "context": "deployment=nginx@0.0.2,support=nginx,app=test",
    },
    {
      "context": "execution=nginx@0.0.2_qa,deployment=nginx@0.0.2,support=nginx,app=test,environment=qa,region=region-1,app=test",
    },
  ],
}
`;
